### ISSUE TITLE
Support plugin shutdown

### ISSUE THREAD

**Issue by AdriVanHoudt** [hapijs/hapi#2389]
The idea is to let plugins gracefully handle a server shutdown. 
If `server.stop()` is called you can't do anything async in your plugin anymore.
Example:
```javascript
server.on('stop', function () {
    internals.pool.end(function (err) {
        if (err) {
            server.log(internals.logTags, 'Failed to gracefully end the pool');
        }

        internals.pool = null;
    });
});
```
The server won't wait for the ending of the pool to close.

**Comment by hueniverse** [comment]
Got an API in mind?

**Comment by AdriVanHoudt** [comment]
Could be 2 things
* Register some form of tail functions for the server
* Pass a callback to the event and wait for all callbacks from all listeners 
    * Or something more blocking like `reply.continue()` where you at some point give control back to the server (although this doesn't seem like a good idea)

**Comment by AdriVanHoudt** [comment]
@hueniverse any thoughts? I would do a PR but I think this goes beyond my knowledge of Hapi internals

**Comment by hueniverse** [comment]
Basically, I need a new event emitter class that allows async listeners. I'll work on it.

**Comment by AdriVanHoudt** [comment]
Cool thanks!

**Comment by Marsup** [comment]
@hueniverse Are you saying you'll work on switching from domains to async-listener ?

**Comment by hueniverse** [comment]
Noooo. Not until it's actually done, and available in the version of node we support...

I meant a new event emitter class that allows setting event listeners where the listener method is async and uses a callback to signal that the next listener handler can be called.

**Comment by akuz** [comment]
Any thoughts on this? https://github.com/opentable/hapi-shutdown

**Comment by AdriVanHoudt** [comment]
@akuz cool plugin but I'd rather have this natively in Hapi

**Comment by akuz** [comment]
@AdriVanHoudt I agree... I was just wondering, if someone taken a look and thought if it's designed well, so I could use something now...

**Comment by Marsup** [comment]
This is weird that server is started before all your plugins are registered, I wouldn't use it personally.

**Comment by hueniverse** [comment]
Resolved via #2685

**Comment by AdriVanHoudt** [comment]
Nice, thanks!
